File: n8n/mcp-validation.js
// Versioned MCP validation shared by the exported writer and reference consumer harness.
const mcpFail = () => { throw new Error('MCP_VALIDATION_FAILED'); };
const mcpAssert = (condition) => { if (!condition) mcpFail(); };
const mcpObject = (value) => value !== null && typeof value === 'object' && !Array.isArray(value);
const mcpStable = (value) => JSON.stringify(value, (_key, v) => mcpObject(v) ? Object.fromEntries(Object.keys(v).sort().map(k => [k,v[k]])) : v);
const mcpDate = value => typeof value === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(value) && Number.isFinite(Date.parse(value)) && new Date(value).toISOString().slice(0,10) === value;
const mcpInstant = value => typeof value === 'string' && /T.*(Z|[+-]\d{2}:\d{2})$/.test(value) && mcpDate(value.slice(0,10)) && Number.isFinite(Date.parse(value));
const mcpSchemaValid = (value, schema, definitions=mcpContract.$defs) => {
  if (!mcpObject(schema)) return false;
  if (schema.$ref && !mcpSchemaValid(value, definitions[schema.$ref.split('/').pop()], definitions)) return false;
  if (schema.const !== undefined && value !== schema.const) return false;
  if (schema.enum && !schema.enum.includes(value)) return false;
  if (schema.anyOf && !schema.anyOf.some(s=>mcpSchemaValid(value,s,definitions))) return false;
  if (schema.oneOf && schema.oneOf.filter(s=>mcpSchemaValid(value,s,definitions)).length !== 1) return false;
  if (schema.allOf && !schema.allOf.every(s=>mcpSchemaValid(value,s,definitions))) return false;
  if (schema.not && mcpSchemaValid(value,schema.not,definitions)) return false;
  if (schema.if) {
    const branch=mcpSchemaValid(value,schema.if,definitions) ? schema.then : schema.else;
    if (branch && !mcpSchemaValid(value,branch,definitions)) return false;
  }
  if (schema.type) {
    const types={object:mcpObject(value),array:Array.isArray(value),string:typeof value==='string',
      boolean:typeof value==='boolean',null:value===null,number:typeof value==='number'&&Number.isFinite(value),
      integer:typeof value==='number'&&Number.isSafeInteger(value)};
    if (!types[schema.type]) return false;
  }
  if (typeof value === 'string') {
    if (schema.minLength !== undefined && [...value].length<schema.minLength) return false;
    if (schema.maxLength !== undefined && [...value].length>schema.maxLength) return false;
    if (schema.pattern && !(new RegExp(schema.pattern,'u')).test(value)) return false;
    if (schema.format==='date' && !mcpDate(value)) return false;
    if (schema.format==='date-time' && !mcpInstant(value)) return false;
    if (schema.format==='uuid' && !/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(value)) return false;
  }
  if (typeof value === 'number') {
    if (schema.minimum !== undefined && value<schema.minimum) return false;
    if (schema.maximum !== undefined && value>schema.maximum) return false;
  }
  if (Array.isArray(value)) {
    if (schema.minItems !== undefined && value.length<schema.minItems) return false;
    if (schema.maxItems !== undefined && value.length>schema.maxItems) return false;
    if (schema.items && !value.every(v=>mcpSchemaValid(v,schema.items,definitions))) return false;
    if (schema.contains && !value.some(v=>mcpSchemaValid(v,schema.contains,definitions))) return false;
    if (schema.uniqueItems && new Set(value.map(mcpStable)).size!==value.length) return false;
  }
  if (mcpObject(value)) {
    if (schema.required && !schema.required.every(k=>Object.hasOwn(value,k)&&value[k]!==undefined)) return false;
    if (schema.properties && !Object.entries(schema.properties).every(([k,s])=>!Object.hasOwn(value,k)||mcpSchemaValid(value[k],s,definitions))) return false;
    if (schema.additionalProperties===false && Object.keys(value).some(k=>!Object.hasOwn(schema.properties||{},k))) return false;
  }
  return value!==undefined;
};
const mcpSchema = (value,name) => mcpAssert(mcpSchemaValid(value,mcpContract.$defs[name]));
const mcpE = value => {
  mcpSchema(value,'id');
  return encodeURIComponent(value).replace(/[!'()*]/g,c=>'%'+c.charCodeAt(0).toString(16).toUpperCase());
};
const mcpKey = (template,values) => mcpContract.identity[template].replace(/\{E\(([^)]+)\)\}/g,(_match,k)=>mcpE(values[k]));
const mcpUnique = (rows,fields) => mcpAssert(new Set(rows.map(row=>mcpStable(fields.map(k=>row[k])))).size===rows.length);
const mcpDecimal = value => {
  mcpSchema(value,'decimal');
  const negative=value.startsWith('-');
  const [whole,fraction='']=value.replace(/^-/,'').split('.');
  return BigInt(whole+fraction.padEnd(18,'0'))*(negative?-1n:1n);
};
const mcpMoney = (money,currency=null) => {
  if (currency!==null) mcpAssert(money.currency===currency);
  if (money.amount===null) { mcpAssert(money.amount_eur===null&&money.eur_basis==='UNAVAILABLE'); return; }
  mcpAssert(money.currency!==null);
  if (money.currency==='EUR') mcpAssert(money.amount_eur!==null&&money.eur_basis==='SOURCE_EUR'&&mcpDecimal(money.amount)===mcpDecimal(money.amount_eur));
  else if (currency!==null) mcpAssert(money.amount_eur===null&&money.eur_basis==='UNAVAILABLE');
};
const mcpParisDate = value => {
  const parts=Object.fromEntries(new Intl.DateTimeFormat('en-CA',{timeZone:'Europe/Paris',year:'numeric',month:'2-digit',day:'2-digit'}).formatToParts(new Date(value)).map(p=>[p.type,p.value]));
  return `${parts.year}-${parts.month}-${parts.day}`;
};
const mcpSnapshot = snapshot => {
  mcpSchema(snapshot,'snapshot_v3');
  const p=snapshot.provenance,c=snapshot.coverage;
  const start=Date.parse(p.collection_started_at),end=Date.parse(p.collection_ended_at),generated=Date.parse(snapshot.generated_at);
  mcpAssert(start<=end&&end<=generated&&end-start<=180000&&snapshot.snapshot_date===mcpParisDate(snapshot.generated_at));
  const keys={accounts:['account_key'],connections:['connection_key'],ownership:['account_key','owner_key'],positions:['holding_type','holding_id'],position_rates:['position_key','source_field'],allocation_categories:['category'],allocation_types:['category','holding_type'],members:['member_ordinal'],warnings:['code','entity_key'],unsupported_details:['account_key','holding_type','reason']};
  for (const [table,fields] of Object.entries(keys)) mcpUnique(snapshot[table],fields);
  const accounts=new Map(snapshot.accounts.map(a=>[a.account_key,a]));
  const connections=new Set(snapshot.connections.map(c=>c.connection_key));
  for (const a of snapshot.accounts) {
    mcpAssert(a.account_key===mcpKey('account_key',{account_id:a.source_account_id}));
    mcpAssert(a.connection_state==='NO_CONNECTION'?a.connection_key===null:a.connection_state==='RESOLVED'?connections.has(a.connection_key):a.connection_key!==null);
    mcpMoney(a.native_balance);
    if(a.native_balance.eur_basis==='SOURCE_CONVERSION')mcpAssert(a.full_value_eur!==null&&mcpDecimal(a.native_balance.amount_eur)===mcpDecimal(a.full_value_eur));
  }
  for(const o of snapshot.ownership)mcpAssert(accounts.get(o.account_key)?.ownership_evidence==='EXPLICIT_ENTRIES'&&o.owner_key===mcpKey('owner_key',{owner_type:o.owner_type,owner_id:o.owner_source_id}));
  for(const row of snapshot.positions){
    const account=accounts.get(row.account_key);mcpAssert(Boolean(account));
    const values={account_id:account.source_account_id,holding_id:row.holding_id,holding_type:row.holding_type};
    mcpAssert(row.source_asset_id===mcpKey('source_asset_id',values)&&row.position_key===mcpKey('position_key',values));
    for(const field of ['current_value','buying_price']){mcpMoney(row[field]);mcpAssert(row[field].eur_basis!=='SOURCE_CONVERSION');}
  }
  const positions=new Set(snapshot.positions.map(p=>p.position_key));
  for(const row of snapshot.position_rates)mcpAssert(positions.has(row.position_key));
  for(const row of snapshot.unsupported_details)mcpAssert(accounts.has(row.account_key));
  const categories=new Set(snapshot.allocation_categories.map(a=>a.category));
  for(const row of snapshot.allocation_types)mcpAssert(categories.has(row.category));
  for(const row of [...snapshot.allocation_categories,...snapshot.allocation_types])mcpMoney(row.value,p.currency);
  for(const group of [snapshot.overview,...snapshot.members]){
    for(const field of ['gross_assets','reported_liabilities','reported_net_worth'])mcpMoney(group[field],p.currency);
    mcpAssert(group.reported_liabilities.amount===null||mcpDecimal(group.reported_liabilities.amount)>=0n);
  }
  mcpMoney(snapshot.overview.financial_assets,p.currency);
  for(const [dimension,rule]of Object.entries(mcpContract.valuation_contracts)){
    const rows=snapshot[rule.rows],known=rows.filter(r=>r[rule.money_field].amount!==null&&r[rule.money_field].currency!==null).length;
    mcpAssert(c[dimension]===(c[rule.collection]!=='COMPLETE'?'UNAVAILABLE':known===rows.length?'COMPLETE':known?'PARTIAL':'UNAVAILABLE'));
  }
  return snapshot;
};
const mcpRun = (run,executionId) => {
  mcpAssert(mcpObject(run)&&typeof executionId==='string'&&/^[A-Za-z0-9_-]{1,128}$/.test(executionId));
  mcpAssert(new RegExp(`^n8n-run:${executionId}:[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`).test(run.run_id));
  mcpAssert(mcpInstant(run.started_at)&&Number.isSafeInteger(run.started_epoch_ms)&&Math.abs(Date.parse(run.started_at)-run.started_epoch_ms)<1000);
  mcpAssert(run.provider==='finary_official_mcp'&&run.api_schema==='3.0'&&run.workbook_schema==='3.0'&&run.source_contract_version==='1.0.0');
  mcpAssert(typeof run.writer_id==='string'&&run.writer_id.length>0&&Number.isSafeInteger(run.writer_generation)&&run.writer_generation>=1);
};
const mcpControl = (rows,run) => {
  mcpAssert(rows.length===1);
  const {row_key,...control}=rows[0];
  mcpSchema(control,'writer_control');
  mcpAssert(row_key==='singleton'&&control.state==='ACTIVE'&&control.writer_id===run.writer_id&&control.generation===run.writer_generation);
};

File: finary-bridge/app/mcp_validation.py
"""Production schema and semantic validation for normalized MCP observations."""

from __future__ import annotations

import json
from datetime import datetime
from decimal import Decimal
from functools import lru_cache
from importlib.resources import files
from typing import Any, ClassVar
from urllib.parse import quote
from zoneinfo import ZoneInfo

from jsonschema import Draft202012Validator, FormatChecker
from pydantic import BaseModel, ConfigDict, model_validator

CONTRACT: dict[str, Any] = json.loads(files("app").joinpath("mcp-contract.json").read_text())
FORMATS = FormatChecker()


@FORMATS.checks("date-time", raises=(TypeError, ValueError))
def aware_timestamp(value: object) -> bool:
    return isinstance(value, str) and datetime.fromisoformat(value).utcoffset() is not None


@lru_cache
def schema_validator(name: str) -> Draft202012Validator:
    return Draft202012Validator(
        {"$ref": f"#/$defs/{name}", "$defs": CONTRACT["$defs"]}, format_checker=FORMATS
    )


def validate(name: str, value: Any) -> None:
    """Never include validation paths, input values or upstream messages in failures."""
    try:
        valid = schema_validator(name).is_valid(value)
    except (ValueError, TypeError, RecursionError):
        valid = False
    if not valid:
        raise ValueError("MCP contract validation failed") from None


def key(template: str, **components: str) -> str:
    for value in components.values():
        validate("id", value)
    return str(
        CONTRACT["identity"][template].format(
            **{f"E({name})": quote(value, safe="-._~") for name, value in components.items()}
        )
    )


def require(condition: bool) -> None:
    if not condition:
        raise ValueError("MCP semantic validation failed")


def unique(rows: list[dict[str, Any]], *columns: str) -> None:
    identities = [tuple(row[column] for column in columns) for row in rows]
    require(len(set(identities)) == len(identities))


def validate_money(value: dict[str, Any], view_currency: str | None = None) -> None:
    amount, eur, currency = value["amount"], value["amount_eur"], value["currency"]
    basis = value["eur_basis"]
    if view_currency is not None:
        require(currency == view_currency)
    if amount is None:
        require(eur is None and basis == "UNAVAILABLE")
        return
    require(currency is not None)
    if currency == "EUR":
        require(eur is not None and basis == "SOURCE_EUR")
        require(Decimal(amount) == Decimal(eur))
    elif view_currency is not None:
        require(eur is None and basis == "UNAVAILABLE")


def validate_position(position: dict[str, Any]) -> None:
    """Validate retained position semantics without borrowing current account data."""
    components = {name: position[name] for name in ("holding_type", "holding_id")}
    require(position["source_asset_id"] == key("source_asset_id", **components))
    template = CONTRACT["identity"]["position_key"].replace(
        CONTRACT["identity"]["account_key"], "{account_key}", 1
    )
    expected = template.format(
        account_key=position["account_key"],
        **{f"E({name})": quote(value, safe="-._~") for name, value in components.items()},
    )
    require(position["position_key"] == expected)
    for field in ("current_value", "buying_price"):
        validate_money(position[field])
        require(position[field]["eur_basis"] != "SOURCE_CONVERSION")


def validate_collection_context(value: dict[str, Any]) -> None:
    """Validate the collection window independently of retained detail availability."""
    p = value["provenance"]
    generated = datetime.fromisoformat(value["generated_at"])
    start, end = (
        datetime.fromisoformat(p[f"collection_{part}_at"]) for part in ("started", "ended")
    )
    require(start <= end <= generated)
    require(
        (end - start).total_seconds()
        <= CONTRACT["transport_policy"]["limits"]["collection_timeout_seconds"]
    )
    require(
        generated.astimezone(ZoneInfo("Europe/Paris")).date().isoformat() == value["snapshot_date"]
    )


def validate_snapshot(value: dict[str, Any]) -> None:
    """Cross-record rules complement the packaged JSON Schema, never detail sums."""
    validate_collection_context(value)
    p = value["provenance"]
    table_keys = {
        "accounts": ("account_key",),
        "connections": ("connection_key",),
        "ownership": ("account_key", "owner_key"),
        "positions": ("holding_type", "holding_id"),
        "position_rates": ("position_key", "source_field"),
        "members": ("member_ordinal",),
        "allocation_categories": ("category",),
        "allocation_types": ("category", "holding_type"),
        "warnings": ("code", "entity_key"),
        "unsupported_details": ("account_key", "holding_type", "reason"),
    }
    for table, columns in table_keys.items():
        unique(value[table], *columns)
    accounts = {row["account_key"]: row for row in value["accounts"]}
    connections = {row["connection_key"] for row in value["connections"]}
    for account in accounts.values():
        require(
            account["account_key"] == key("account_key", account_id=account["source_account_id"])
        )
        connection, state = account["connection_key"], account["connection_state"]
        require(
            (state != "NO_CONNECTION" or connection is None)
            and (state != "UNRESOLVED" or connection is not None)
            and (state != "RESOLVED" or connection in connections)
        )
        validate_money(account["native_balance"])
        if account["native_balance"]["eur_basis"] == "SOURCE_CONVERSION":
            require(account["full_value_eur"] is not None)
            require(
                Decimal(account["native_balance"]["amount_eur"])
                == Decimal(account["full_value_eur"])
            )
    for owner in value["ownership"]:
        require(owner["account_key"] in accounts)
        require(accounts[owner["account_key"]]["ownership_evidence"] == "EXPLICIT_ENTRIES")
        require(
            owner["owner_key"]
            == key("owner_key", owner_type=owner["owner_type"], owner_id=owner["owner_source_id"])
        )
    positions = set()
    for position in value["positions"]:
        require(position["account_key"] in accounts)
        validate_position(position)
        positions.add(position["position_key"])
    for rate in value["position_rates"]:
        require(rate["position_key"] in positions)
    for diagnostic in value["unsupported_details"]:
        require(diagnostic["account_key"] in accounts)
    for dimension, rule in CONTRACT["valuation_contracts"].items():
        rows = value[rule["rows"]]
        known = sum(
            row[rule["money_field"]]["amount"] is not None
            and row[rule["money_field"]]["currency"] is not None
            for row in rows
        )
        expected = "UNAVAILABLE"
        if value["coverage"][rule["collection"]] == "COMPLETE":
            expected = "COMPLETE" if known == len(rows) else "PARTIAL" if known else "UNAVAILABLE"
        require(value["coverage"][dimension] == expected)
    for group in [value["overview"], *value["members"]]:
        for name in ("gross_assets", "reported_liabilities", "reported_net_worth"):
            validate_money(group[name], p["currency"])
        debt = group["reported_liabilities"]["amount"]
        require(debt is None or Decimal(debt) >= 0)
    validate_money(value["overview"]["financial_assets"], p["currency"])
    categories = {row["category"] for row in value["allocation_categories"]}
    for allocation in value["allocation_types"]:
        require(allocation["category"] in categories)
    for allocation in value["allocation_categories"] + value["allocation_types"]:
        validate_money(allocation["value"], p["currency"])


class ContractModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, strict=True, hide_input_in_errors=True)
    contract_name: ClassVar[str]

    @model_validator(mode="before")
    @classmethod
    def check_contract(cls, value: Any) -> Any:
        if isinstance(value, cls):
            return value
        validate(cls.contract_name, value)
        if cls.contract_name == "snapshot_v3":
            validate_snapshot(value)
        elif cls.contract_name == "money":
            validate_money(value)
        return value
